crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -38,8 +38,35 @@ impl RemovedSystem {
 ///
 /// These are opaque identifiers, keyed to a specific [`World`],
 /// and are created via [`World::register_system`].
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub struct SystemId(Entity);
+#[derive(Eq)]
+pub struct SystemId<I = (), O = ()>(Entity, std::marker::PhantomData<fn(I) -> O>);
```

Comment:
The particular type in the `PhantomData` doesn't really matter as long as it mentions both `I` and `O`, so I figured it might as well be `fn(I) -> O` since that (mnemonically) describes how the parameters are used.